src/hotspot/share/classfile/systemDictionary.cpp

```diff
@@ -1490,16 +1493,19 @@ void SystemDictionary::define_instance_class(InstanceKlass* k, Handle class_load
   }
 
   // Add the new class. We need recompile lock during update of CHA.
+  DeoptimizationScope deopt_scope;
   {
     MutexLocker mu_r(THREAD, Compile_lock);
 
     // Add to class hierarchy, and do possible deoptimizations.
-    add_to_hierarchy(k);
+    add_to_hierarchy(&deopt_scope, k);
 
     // Add to systemDictionary - so other classes can see it.
     // Grabs and releases SystemDictionary_lock
     update_dictionary(THREAD, k, loader_data);
```

Comment:
All these patterns are the same in class loading, except here we update the SystemDictionary after deoptimizing holding the Compile_lock and in your change we update it beforehand.
I was going to suggest future cleanup of moving add_to_hierarchy and the associated deopt_scope code to InstanceKlass, but now I'm not sure if this order matters.
Holding the Compile_lock while adding a class to the SystemDictionary and the code to hold the Compile_lock in ciEnv::get_klass_by_name() now doesn't make sense to me.  Klass is read lock free in the dictionary and temporarily holding the Compile_lock doesn't seem to do what it says:
``` 
{ // Grabbing the Compile_lock prevents systemDictionary updates
  // during compilations.
```

At any rate, I think the SystemDictionary might need to be updated after deoptimization is done.  Can you check this?

Reply:
Please see new commit, thanks!